src/member.ts

```diff
@@ -0,0 +1,358 @@
+
+  constructor(req: Request) {
+    this.req = req;
+    this.signatures = {};
+    this.finished = false;
+  }
+
+  getState(
+    neededSignatures: number,
+    maxDenied: number = neededSignatures
+  ): ResponseState {
+    if (this.finished) return "finished";
+    if (this.isCancelled()) return "cancelled";
+    if (this.isConflicted()) return "failed";
+    if (this.numberAccepted() >= neededSignatures) return "ready";
+    if (this.numberDenied() >= maxDenied) return "failed";
+
+    return "pending";
+  }
+
+  addResponse(author: ID, response: Response) {
+    this.signatures[author] = response;
+  }
+
+  get id() {
+    return this.req.id;
+  }
+
+  get from() {
+    return this.req.from;
+  }
+
+  get operation() {
+    return this.req.operation;
+  }
+
+  numberDenied(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "deny";
+    }).length;
+  }
+
+  numberAccepted(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "accept";
+    }).length;
+  }
+
+  isConflicted(): boolean {
+    return Object.keys(this.signatures).some((id: ID) => {
+      return this.signatures[id].response === "conflict";
+    });
+  }
+
+  isCancelled(): boolean {
+    return this.signatures[this.from]?.response === ("cancel" as ResponseType);
+  }
+
+  finish() {
+    this.finished = true;
+  }
+}
+
+function allowAll(request: Request): boolean {
+  return true;
+}
+
+export interface ShouldAcceptCB {
+  (request: Request): boolean;
+}
+
+export interface MemberConstructorOptions {
+  id?: ID;
+  initiator?: ID;
+  shouldAccept?: ShouldAcceptCB;
+}
+
+export class Member extends EventEmitter {
+  id: ID;
+  ownFeed: FeedItem[];
+  knownFeeds: {
+    [id in ID]: FeedItem[];
+  };
+  memberIndexes: {
+    [id in ID]: number;
+  };
+  ownIndex: number;
+  requests: {
+    [id in ID]: RequestState;
+  };
+  knownMembers: string[];
+  shouldAccept: ShouldAcceptCB;
+
+  constructor(
+    { id, initiator, shouldAccept = allowAll }: MemberConstructorOptions = {
+      shouldAccept: allowAll,
+    }
+  ) {
+    super();
+    this.id = id || makeID();
+    this.ownFeed = [];
+    this.ownIndex = 0;
+    this.knownFeeds = {
+      [this.id]: this.ownFeed,
+    };
+    this.memberIndexes = {
+      [this.id]: 0,
+    };
+    this.requests = {};
+    this.knownMembers = [initiator || this.id];
+    this.shouldAccept = shouldAccept;
+  }
+
+  sync(member: Member) {
+    // Iterate through member's knownFeeds
+    // Get any new data you don't have
+    // Get new data from members to knownFeeds
+    for (const id of Object.keys(member.knownFeeds)) {
+      const feed = member.knownFeeds[id];
+      const ownCopy = this.getFeedFor(id);
+      if (ownCopy.length < feed.length) {
+        this.updateFeedFor(id, feed);
+      }
+    }
+  }
+
+  processFeeds(): boolean {
+    let hasProcessed = false;
+    let hasResponded = false;
```

Comment:
`hasResponded` is set but not used.

Reply:
Is there a way to get typescript to detect stuff like that? I feel like I might have to add eslint into the mix.

Having three tools to do linting feels a bit daunting. 😂 Parsing all the code at least 3 times per lint.

Reply:
The variable was supposed to be used later. Good catch!

Reply:
What do you need Typescript for to detect this if you have Martin? ;-)

Reply:
Note: my IDE uses code colors to highlight unused variables.

Reply:
It'd be nice if the typescript type checking was inside eslint, too. I'll see if that's an option. 😁